metricbeat/module/kafka/broker.go

```diff
@@ -290,6 +303,114 @@ func (b *Broker) FetchPartitionOffsetFromTheLeader(topic string, partitionID int
 	return offset, nil
 }
 
+// FetchPartitionOffsetsForTopics fetches offsets of all partitions from its leaders.
+func (b *Broker) FetchPartitionOffsetsForTopics(topics []*sarama.TopicMetadata, time int64) map[string]map[int32]PartitionOffsets {
+	leaderTopicPartition := b.groupBrokersPerTopicPartitions(topics)
+	topicPartitionPartitionOffsets := b.fetchGroupedPartitionOffsetsPerBroker(leaderTopicPartition, time)
+	return topicPartitionPartitionOffsets
+}
+
+func (b *Broker) groupBrokersPerTopicPartitions(topics []*sarama.TopicMetadata) brokerTopicPartitionsMap {
+	leaderTopicPartition := brokerTopicPartitionsMap{}
+	for _, topic := range topics {
+		for _, partition := range topic.Partitions {
+			if _, ok := leaderTopicPartition[partition.Leader]; !ok {
+				leaderTopicPartition[partition.Leader] = []topicPartition{}
+			}
+			leaderTopicPartition[partition.Leader] = append(leaderTopicPartition[partition.Leader],
+				topicPartition{topic: topic.Name, partition: partition.ID})
+		}
+	}
+	return leaderTopicPartition
+}
+
+func (b *Broker) fetchGroupedPartitionOffsetsPerBroker(leaderTopicPartition brokerTopicPartitionsMap,
+	time int64) map[string]map[int32]PartitionOffsets {
+
+	var wg sync.WaitGroup
+	wg.Add(len(leaderTopicPartition))
+
+	queryResults := make(chan map[string]map[int32]PartitionOffsets, len(leaderTopicPartition))
+	defer close(queryResults)
+
+	for leader, topicPartitions := range leaderTopicPartition {
+		thisLeader := leader
+		thisTopicPartitions := topicPartitions
+		go func() {
+			queryResults <- b.queryBrokerForPartitionOffsets(thisLeader, thisTopicPartitions, time)
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+
+	topicPartitionPartitionOffsets := map[string]map[int32]PartitionOffsets{}
+	for i := 0; i < len(leaderTopicPartition); i++ {
+		queryResult := <-queryResults
+		for topic := range queryResult {
+			topicPartitionPartitionOffsets[topic] = queryResult[topic]
+		}
+	}
+	return topicPartitionPartitionOffsets
+}
+
+func (b *Broker) queryBrokerForPartitionOffsets(brokerID int32, topicPartitions []topicPartition, time int64) map[string]map[int32]PartitionOffsets {
+	req := new(sarama.OffsetRequest)
```

Comment:
Something I am missing here is that in previous implementation we were making a request per replica (the request done in `b.PartitionOffset()` was also modified with `req.SetReplicaID()`).
Are we missing the offsets of replicas now? :thinking:

Reply:
It seems that this may work (replicaID = -1 means a leader), with I believe this is getting too complex.